README.md
# Download at
- https://mods.factorio.com/mod/Armored-train-With-Inserters

### Fork of
- https://mods.factorio.com/mod/Armored-train

### What's inside

* Armored locomotive (Crush tough biters)
* Minigun platform MK1 (Shoot at close range)
* Rocket platform MK1 (Shoot at long range, can accept any rockets and nuke)
* Cannon wagon MK1 (Shoot at medium range, can accept any cannon ammo)
* Flamethrower wagon MK1 (Shoot at long range, fire damage)
* Radar platform (Reveal train on map)

### Changes to the original
- Armoured Waggons can now get ammo via inserters and have a bigger inventory



changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Info:
	- This is the point at which i forked the project
	- The waggons now have an inventory. They will give the ammo to the turret

---------------------------------------------------------------------------------------------------
Version: 0.4.6
Info:
	- Thumbnail fix

---------------------------------------------------------------------------------------------------
Version: 0.4.5
Info:
	- Changed sprite for radar (Limited Distribution Only Licence of original mod by Earendel)

---------------------------------------------------------------------------------------------------
Version: 0.4.4
Info:
	- Changelog fix!
	- Dependencies fix!
	- Title fix!

---------------------------------------------------------------------------------------------------
Version: 0.4.3
Info:
	- Update for 1.1 factorio (experimental)

---------------------------------------------------------------------------------------------------
Version: 0.4.2
Info:
	- Radar platform! (fuel powered)
	- Icons improved!
	- Recipes balanced!
	- Localization (RU, KO auto-translated)

---------------------------------------------------------------------------------------------------
Version: 0.4.1
Info:
	- Flamethrower turret!
	- Turret parameters balanced again

---------------------------------------------------------------------------------------------------
Version: 0.4.0
Info:
	- Cannon turret! (Changes vanilla tank ammo - not to damage player forces)
	- Projectile offset fixes
	- Bug when can not exit wagon/platform fixed
	- Turret parameters balanced

---------------------------------------------------------------------------------------------------
Version: 0.3.13 BETA
Info:
	- Wagon turrets re-implimented (optimised, can accept all types of ammo)
	- Removed unnecessary items and recipes
	- Cannon turret in future updates (need to fix sprite offsets)
	- No migrations, because everything was re-implemented!

---------------------------------------------------------------------------------------------------
Version: 0.3.12
Info:
	- Update for 1.0

---------------------------------------------------------------------------------------------------
Version: 0.3.11
Info:
	- Changelog fix

---------------------------------------------------------------------------------------------------
Version: 0.3.10
Info:
	- Few scripts changes for compatibility with some mods...

---------------------------------------------------------------------------------------------------
Version: 0.3.0
Info:
	- Mod portal version bug fix.

---------------------------------------------------------------------------------------------------
Version: 0.2.2
Info:
	- Fixed mask and shadow for locomotive.
	- Turret range tweaked.
	- Minor fixes.

---------------------------------------------------------------------------------------------------
Version: 0.2.16 
Info:
	- Fixed shadows and masks for some wagons.
	- Minor fixes.
	- P.S. A. Train will be remade in future update. B. Ammo compatibility coming in next update.

---------------------------------------------------------------------------------------------------
Version: 0.2.151 BETA
Info:
	- Fixed graphics for rocket, cannon, minigun platforms/wagons.
	- Fix for factorio 0.17.34 (See fff #293)

Changes:
	- Cannon platform are no more! Replaced with Cannon Wagon!
	- Other minor fixes

---------------------------------------------------------------------------------------------------
Version: 0.2.11 BETA
Info:
	- Added rocket turret!

Changes:
	- Speed force of armored locomotive fixed!
	- Tech and upgrades are fixed!
	- Minor fixes.

---------------------------------------------------------------------------------------------------
Version: 0.2.0
Info:
	- This is a BETA version.

Changes:
	- Now there are platforms with turrets/cannons/ etc. In the future they will be modular, you build 1 platform and put (I mean craft) on them whatever you want, even a radar.
	- Graphics a little bit "strange" (blame Factorio's sprite render-layers), but still looking good.
	- Added more bugs and glitches.

---------------------------------------------------------------------------------------------------
Version: 0.1.52
Changes:
	- Fixed Hp and other minor values

---------------------------------------------------------------------------------------------------
Version: 0.1.51
Info:
	- 0.17. version of game compatibility

---------------------------------------------------------------------------------------------------
Version: 0.1.2
Bugfixes:
	- Fixed issue with Vehicle Turret mod, now can run both. (Additional testing required)
	- Fixed technology tree and descriptions.
	- Minor fixes.

Changes:
	- Migrations are written so no broken saves, and correct map loading.
	- Removed recipe for rocket turrets. (They are still in game, will add them in future updates as soon as I fix them and add turret cannons)
	- Added new bugs. (Report if any stuff is broken)

---------------------------------------------------------------------------------------------------
Version: 0.1.11
Bugfixes:
	- Fixed displayed names.

Changes:
	- Now able to install turret only in minigun wagon, not vanilla one. The vanilla one is working like a storage as it should be.
	- Expanded selection box so player can repair his turrets.

---------------------------------------------------------------------------------------------------
Version: 0.1.1
Bugfixes:
	- Fixed sprite bug on startup.

---------------------------------------------------------------------------------------------------
Version: 0.1.0
Info:
	- Initial release.

control.lua
--------------------------
--FUNCTIONS---------------
--------------------------
--Is this mod entity? If yes true, othervise false
isModEntity = function(modEntity)
	if modEntity and modEntity.name ~= nil then		-- null fix
		--List of known entities
		if modEntity.name == "minigun-platform-mk1" 
			or modEntity.name == "rocket-platform-mk1" 
			or modEntity.name == "cannon-wagon-mk1" 
			or modEntity.name == "flamethrower-wagon-mk1"
			or modEntity.name == "radar-platform-mk1" then	
			return true
		end
	end
	return false
end

--Create turret as proxy to platform (data stored to table)
function createProxyTurretMinigun(position, surface, force)
	local proxy = surface.create_entity{
		name = "minigun-turret-mk1", 
		position = position, 
		force = force
	}
	return proxy
end

function createProxyTurretRocket(position, surface, force)
	local proxy = surface.create_entity{
		name = "rocket-turret-mk1", 
		position = position, 
		force = force
	}
	return proxy
end

function createProxyTurretCannon(position, surface, force)
	local proxy = surface.create_entity{
		name = "cannon-turret-mk1", 
		position = position, 
		force = force
	}
	return proxy
end

function createProxyTurretFlamethrower(position, surface, force)
	local proxy = surface.create_entity{
		name = "flamethrower-turret-mk1", 
		position = position, 
		force = force
	}
	return proxy
end

function createProxyRadar(position, surface, force)
	local proxy = surface.create_entity{
		name = "radar-mk1", 
		position = position, 
		force = force
	}
	return proxy
end

--Create initTurretPlatformList table to store data for proxy and platform or pass data if created
function initTurretPlatformList(tableValue)
	--If first time calling = create table?
	if tableValue == nil then
		return {}
	--if table created just pass values
	else
		return tableValue
	end
end

-------------
--ON_EVENTS--
-------------

--ON LOAD \/
--Prevent on load bug when no hp ws found
script.on_load(function(event)
	platformMaxHealth = 1000
end)
--ON LOAD /\


--ON BUILT \/--
function entityBuilt(event)
	--createdEntity reference (to simplify usage in this context)
	local createdEntity = {}
    if event.name == defines.events.script_raised_built then 
        createdEntity = event.entity
    else
        createdEntity = event.created_entity
    end
	
	--Is this mod entity?
	if isModEntity(createdEntity) then
		--createdPlatform now defines itself with created platform + turret (later on)
		local createdPlatform = {
			--Actual entity as class
			entity = createdEntity
		}
		
		if createdEntity.name == "minigun-platform-mk1" then
			--Create and define a proxy at the created entity position, surface, and force
			createdPlatform.proxy = createProxyTurretMinigun(
				createdEntity.position, 
				createdEntity.surface, 
				createdEntity.force
			)
		end 
		
		if createdEntity.name == "cannon-wagon-mk1" then		
			createdPlatform.proxy = createProxyTurretCannon(
				createdEntity.position, 
				createdEntity.surface, 
				createdEntity.force
			)
		end 
		
		if createdEntity.name == "rocket-platform-mk1" then		
			createdPlatform.proxy = createProxyTurretRocket(
				createdEntity.position, 
				createdEntity.surface, 
				createdEntity.force
			)
		end 
		
		if createdEntity.name == "flamethrower-wagon-mk1" then		
			createdPlatform.proxy = createProxyTurretFlamethrower(
				createdEntity.position, 
				createdEntity.surface, 
				createdEntity.force
			)
		end 
		
		if createdEntity.name == "radar-platform-mk1" then	
			createdPlatform.proxy = createProxyRadar(
				createdEntity.position, 
				createdEntity.surface, 
				createdEntity.force
			)
		end 
		
		--Create table "turretPlatformList" and store data (if null create, else just pass data)
		global.turretPlatformList = initTurretPlatformList(global.turretPlatformList)
		--Add created platform and turret to the table (list)
		table.insert(global.turretPlatformList, createdPlatform)
		
		--Define max health var (for reference)
		platformMaxHealth = 1000
	end
end

--ON_BUILT EVENT
script.on_event(defines.events.on_built_entity, entityBuilt)
script.on_event(defines.events.on_robot_built_entity, entityBuilt)
script.on_event(defines.events.script_raised_built, entityBuilt)
--ON BUILT /\--

--ON TICK \/--
-- Move each turret to follow its wagon and transfer ammo
function onTickMain(event)
    -- If turretPlatformList is not empty (turret and platform placed in world)
    if global.turretPlatformList ~= nil then
        -- For each individual createdPlatform
        for i, createdPlatform in ipairs(global.turretPlatformList) do
            -- Is this entity valid/nil?
            if createdPlatform.proxy ~= nil and createdPlatform.proxy.valid then
                -- Teleport the turret to the platform's position
                createdPlatform.proxy.teleport({
                    x = createdPlatform.entity.position.x,
                    y = createdPlatform.entity.position.y
                })

                -- Transfer ammo from platform to turret
                local cargoWagonInventory = createdPlatform.entity.get_inventory(defines.inventory.cargo_wagon)
                local turretInventory = createdPlatform.proxy.get_inventory(defines.inventory.turret_ammo)

                -- Iterate over the contents of the cargo wagon inventory
                for i = 1, #cargoWagonInventory do
                    local stack = cargoWagonInventory[i]
                    if stack.valid_for_read and turretInventory.can_insert(stack) then
                        -- Transfer the stack from the cargo wagon to the turret
                        local inserted_count = turretInventory.insert(stack)
                        stack.count = stack.count - inserted_count
                        if stack.count == 0 then
                            cargoWagonInventory[i].clear()
                        end
                    end
                end

                -- Taken damage to TURRET is applied to WAGON
                if event.tick % 20 == 3 then
                    local damageTaken = platformMaxHealth - createdPlatform.proxy.health
                    if damageTaken > 0 then
                        local platformCurrentHealth = createdPlatform.entity.health
                        if platformCurrentHealth <= damageTaken then
                            createdPlatform.proxy.destroy()
                            createdPlatform.entity.die()
                        else
                            createdPlatform.entity.health = platformCurrentHealth - damageTaken
                            createdPlatform.proxy.health = platformMaxHealth
                        end
                    end
                end
            end
        end
    end
end

script.on_event(defines.events.on_tick, onTickMain)
--ON TICK /\--

--ON PRE MINED (Remove turret when object not mined but going to be mined) \/--
-- If removed/destroyed
function entityRemoved(event)
	-- Is this a known entity?
	if isModEntity(event.entity) then
		local newFunction = function (val) 
			return val.entity == event.entity
		end
		
		local wagon = getWagonFromEntity(global.turretPlatformList, event.entity)
		
		-- If the wagon is still there
		if wagon ~= nil then
			if wagon.proxy ~= nil and wagon.proxy.valid then
				wagon.proxy.destroy()
				wagon.proxy = nil
			end
			
			-- Remove from the table
			global.turretPlatformList = nilIfEmptyTable(remove_if(newFunction, global.turretPlatformList))
		end
	end
end

script.on_event(defines.events.on_pre_player_mined_item, entityRemoved)
script.on_event(defines.events.on_robot_pre_mined, entityRemoved)
script.on_event(defines.events.script_raised_destroy, entityRemoved)
--ON PRE MINED (Remove turret when object not mined but going to be mined) /\--

--ON MINED (BUFFER for item) \/--
function entityMined(event)
	-- Is this a mod entity?
	if isModEntity(event.entity) then
		-- Define buffer for removed entity
		local bufferRemovedEntity = event.buffer
		
		-- Additional code here if needed
	end
end

script.on_event(defines.events.on_player_mined_entity, entityMined)
--ON MINED (BUFFER) /\--

function entityDestroyed(event)
	-- Is this a known entity?
	if isModEntity(event.entity) then
		local newFunction = function (val) 
			return val.entity == event.entity
		end
		
		local wagon = getWagonFromEntity(global.turretPlatformList, event.entity)
		
		-- If the wagon is still there
		if wagon ~= nil then
			if wagon.proxy ~= nil and wagon.proxy.valid then
				-- Add explosion
				wagon.proxy.destroy()
				-- Alternatively, you can use wagon.damage(10000, game.forces.enemy) instead of destroy
				wagon.proxy = nil
			end
			
			global.turretPlatformList = nilIfEmptyTable(remove_if(newFunction, global.turretPlatformList))
		end
	end
end

script.on_event(defines.events.on_entity_died, entityDestroyed)

function getWagonFromEntity(wagons, entity)
	if wagons == nil then return nil end
	
	for i, value in ipairs(wagons) do
		if isEntityValid(value.entity) and entity == value.entity then
			return value
		end
	end
end

-- Is this entity valid? (not null and valid)
function isEntityValid(validEntity)
	if validEntity ~= nil and validEntity.valid then
		return true
	end
	return false
end

function nilIfEmptyTable(value)
	if value == nil or #value < 1 then
		return nil
	else 
		return value
	end
end

function remove_if(func, arr)
	if arr == nil then return nil end
	local new_array = {}
	for _, v in ipairs(arr) do
		if not func(v) then table.insert(new_array, v) end
	end
	return new_array
end


assets/cannon-wagon-mk1/cannon-wagon-mk1.lua
------------
--- ITEM ---
------------
data:extend
({
	{
		type = "item",
		name = "cannon-wagon-mk1",
		icon = "__Armored-train__/assets/cannon-wagon-mk1/cannon-wagon-mk1-icon.png",
		icon_size = 64, icon_mipmaps = 4,
		subgroup = "transport",
		order = "a[train-system]-l[wagon_mk1_00]",
		place_result = "cannon-wagon-mk1",
		stack_size = 5
	}
})
	
	
	
--------------
--- RECIPE ---
--------------
data:extend
({
	{
		type = "recipe",
		name = "cannon-wagon-mk1",
		enabled = false,						-- default state (not researched)
		ingredients =
		{
			{"iron-plate", 40},			
			{"steel-plate", 40},
			{"iron-gear-wheel", 20},				
			{"engine-unit", 15},
			{"advanced-circuit", 5}			-- Unique
		},
		result = "cannon-wagon-mk1"
	}
})



--------------
--- ENTITY ---
--------------
-- Deep copy base data and create new one with custom parametres
local l_minigun_platform_mk1 = util.table.deepcopy(data.raw["cargo-wagon"]["cargo-wagon"])
l_minigun_platform_mk1.name = "cannon-wagon-mk1"
l_minigun_platform_mk1.icon = "__Armored-train__/assets/cannon-wagon-mk1/cannon-wagon-mk1-icon.png"
l_minigun_platform_mk1.icon_size = 64
l_minigun_platform_mk1.icon_mipmaps = 4
-- Inventory
l_minigun_platform_mk1.inventory_size = 10					
l_minigun_platform_mk1.minable = 
{
	mining_time = 0.5, 
	result = "cannon-wagon-mk1"		
}
l_minigun_platform_mk1.max_health = 1000					-- Same as turret hp (Critical component or will be destroyed using old script)	
l_minigun_platform_mk1.weight = 2000
l_minigun_platform_mk1.energy_per_hit_point = 5
-- Animation
l_minigun_platform_mk1.pictures =
{
	layers =
	{
		{
			priority = "very-low",
			width = 256,
			height = 256,
			back_equals_front = false,    			--means to not rotate mirroring
			direction_count = 128, 					--means to add aditional frames
			allow_low_quality_rotation = true,
			line_length = 32,
			lines_per_file = 1,
			shift = {0.4, -1.25},
			filenames =
			{
				"__Armored-train__/assets/cannon-wagon-mk1/sprites/wagon-mk1-01.png",
				"__Armored-train__/assets/cannon-wagon-mk1/sprites/wagon-mk1-02.png",
				"__Armored-train__/assets/cannon-wagon-mk1/sprites/wagon-mk1-03.png",
				"__Armored-train__/assets/cannon-wagon-mk1/sprites/wagon-mk1-04.png"
			}
		}
	}
}
l_minigun_platform_mk1.horizontal_doors = nil
l_minigun_platform_mk1.vertical_doors = nil
-- Minimap representation
l_minigun_platform_mk1.minimap_representation =
{
	filename = "__Armored-train__/assets/cannon-wagon-mk1/cannon-wagon-mk1-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}
l_minigun_platform_mk1.selected_minimap_representation =
{
	filename = "__Armored-train__/assets/cannon-wagon-mk1/cannon-wagon-mk1-selected-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}

-- Write result
data:extend
({
	l_minigun_platform_mk1
})

assets/radar-platform-mk1/radar-platform-mk1.lua
------------
--- ITEM ---
------------
data:extend
({
	{
		type = "item",
		name = "radar-platform-mk1",
		icon = "__Armored-train__/assets/radar-platform-mk1/radar-platform-mk1-icon.png",
		icon_size = 64, icon_mipmaps = 4,
		subgroup = "transport",
		order = "a[train-system]-l[utility_mk1_00",
		place_result = "radar-platform-mk1",
		stack_size = 5
	}
})
	
	
	
--------------
--- RECIPE ---
--------------
data:extend
({
	{
		type = "recipe",
		name = "radar-platform-mk1",
		enabled = false,							-- default state (not researched)
		ingredients =
		{
			{"iron-plate", 20},			
			{"steel-plate", 20},
			{"iron-gear-wheel", 10},				
			{"engine-unit", 15},
			{"electronic-circuit", 5}			-- Unique
		},
		result = "radar-platform-mk1"
	}
})



--------------
--- ENTITY ---
--------------
-- Deep copy base data and create new one with custom parametres
local l_radar_platform_mk1 = util.table.deepcopy(data.raw["cargo-wagon"]["cargo-wagon"])
l_radar_platform_mk1.name = "radar-platform-mk1"
l_radar_platform_mk1.icon = "__Armored-train__/assets/radar-platform-mk1/radar-platform-mk1-icon.png"
l_radar_platform_mk1.icon_size = 64
l_radar_platform_mk1.icon_mipmaps = 4
-- Inventory
l_radar_platform_mk1.inventory_size = 10
l_radar_platform_mk1.minable = 
{
	mining_time = 0.5, 
	result = "radar-platform-mk1"		
}
l_radar_platform_mk1.max_health = 1000					-- Same as turret hp (Critical component or will be destroyed using old script)	
l_radar_platform_mk1.weight = 2000
l_radar_platform_mk1.energy_per_hit_point = 5
-- Animation
l_radar_platform_mk1.pictures =
{
	layers =
	{
		{
			priority = "very-low",
			width = 256,
			height = 256,
			back_equals_front = false,    			--means to not rotate mirroring
			direction_count = 128, 					--means to add aditional frames
			allow_low_quality_rotation = true,
			line_length = 32,
			lines_per_file = 1,
			shift = {0.4, -1.25},
			filenames =
			{
				"__Armored-train__/assets/radar-platform-mk1/sprites/platform-mk1-01.png",
				"__Armored-train__/assets/radar-platform-mk1/sprites/platform-mk1-02.png",
				"__Armored-train__/assets/radar-platform-mk1/sprites/platform-mk1-03.png",
				"__Armored-train__/assets/radar-platform-mk1/sprites/platform-mk1-04.png"
			}
		},
		--DECORATIVES
		{
			priority = "very-low",
			width = 96,--72
			height = 140,--57
			back_equals_front = false,    		--means to not rotate mirroring
			direction_count = 128, 				--means to add aditional frames
			allow_low_quality_rotation = true,
			line_length = 32,
			lines_per_file = 1,
			--shift = {0.0, -2.2},
			shift = {0.2, -1.35},
			scale = 0.675,
			filenames =
			{
				"__Armored-train__/assets/radar-mk1/sprites/radar-mk1-01-fix.png",
				"__Armored-train__/assets/radar-mk1/sprites/radar-mk1-02-fix.png",
				"__Armored-train__/assets/radar-mk1/sprites/radar-mk1-03-fix.png",
				"__Armored-train__/assets/radar-mk1/sprites/radar-mk1-04-fix.png"
			}
		}
	}
}
l_radar_platform_mk1.horizontal_doors = nil
l_radar_platform_mk1.vertical_doors = nil
-- Minimap representation
l_radar_platform_mk1.minimap_representation =
{
	filename = "__Armored-train__/assets/radar-platform-mk1/radar-platform-mk1-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}
l_radar_platform_mk1.selected_minimap_representation =
{
	filename = "__Armored-train__/assets/radar-platform-mk1/radar-platform-mk1-selected-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}

-- Write result
data:extend
({
	l_radar_platform_mk1
})

assets/rocket-platform-mk1/rocket-platform-mk1.lua
------------
--- ITEM ---
------------
data:extend
({
	{
		type = "item",
		name = "rocket-platform-mk1",
		icon = "__Armored-train__/assets/rocket-platform-mk1/rocket-platform-mk1-icon.png",
		icon_size = 64, icon_mipmaps = 4,
		subgroup = "transport",
		order = "a[train-system]-l[platform_mk1_01]",
		place_result = "rocket-platform-mk1",
		stack_size = 5
	}
})
	
	
	
--------------
--- RECIPE ---
--------------
data:extend
({
	{
		type = "recipe",
		name = "rocket-platform-mk1",
		enabled = false,							-- default state (not researched)
		ingredients =
		{
			{"iron-plate", 20},			
			{"steel-plate", 20},
			{"iron-gear-wheel", 10},				
			{"engine-unit", 15},
			{"electronic-circuit", 20}			-- Unique
		},
		result = "rocket-platform-mk1"
	}
})



--------------
--- ENTITY ---
--------------
-- Deep copy base data and create new one with custom parametres
local l_rocket_platform_mk1 = util.table.deepcopy(data.raw["cargo-wagon"]["cargo-wagon"])
l_rocket_platform_mk1.name = "rocket-platform-mk1"
l_rocket_platform_mk1.icon = "__Armored-train__/assets/rocket-platform-mk1/rocket-platform-mk1-icon.png"
l_rocket_platform_mk1.icon_size = 64
l_rocket_platform_mk1.icon_mipmaps = 4
-- Inventory
l_rocket_platform_mk1.inventory_size = 10					
l_rocket_platform_mk1.minable = 
{
	mining_time = 0.5, 
	result = "rocket-platform-mk1"		
}
l_rocket_platform_mk1.max_health = 1000					-- Same as turret hp (Critical component or will be destroyed using old script)	
l_rocket_platform_mk1.weight = 2000
l_rocket_platform_mk1.energy_per_hit_point = 5
-- Animation
l_rocket_platform_mk1.pictures =
{
	layers =
	{
		{
			priority = "very-low",
			width = 256,
			height = 256,
			back_equals_front = false,    			--means to not rotate mirroring
			direction_count = 128, 					--means to add aditional frames
			allow_low_quality_rotation = true,
			line_length = 32,
			lines_per_file = 1,
			shift = {0.4, -1.25},
			filenames =
			{
				"__Armored-train__/assets/rocket-platform-mk1/sprites/platform-mk1-01.png",
				"__Armored-train__/assets/rocket-platform-mk1/sprites/platform-mk1-02.png",
				"__Armored-train__/assets/rocket-platform-mk1/sprites/platform-mk1-03.png",
				"__Armored-train__/assets/rocket-platform-mk1/sprites/platform-mk1-04.png"
			}
		}
	}
}
l_rocket_platform_mk1.horizontal_doors = nil
l_rocket_platform_mk1.vertical_doors = nil

-- Minimap representation
l_rocket_platform_mk1.minimap_representation =
{
	filename = "__Armored-train__/assets/rocket-platform-mk1/rocket-platform-mk1-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}
l_rocket_platform_mk1.selected_minimap_representation =
{
	filename = "__Armored-train__/assets/rocket-platform-mk1/rocket-platform-mk1-selected-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}

-- Write result
data:extend
({
	l_rocket_platform_mk1
})

assets/minigun-platform-mk1/minigun-platform-mk1.lua
------------
--- ITEM ---
------------
data:extend
({
	{
		type = "item",
		name = "minigun-platform-mk1",
		icon = "__Armored-train__/assets/minigun-platform-mk1/minigun-platform-mk1-icon.png",
		icon_size = 64, icon_mipmaps = 4,
		subgroup = "transport",
		order = "a[train-system]-l[platform_mk1_00]",
		place_result = "minigun-platform-mk1",
		stack_size = 5
	}
})
	
	
	
--------------
--- RECIPE ---
--------------
data:extend
({
	{
		type = "recipe",
		name = "minigun-platform-mk1",
		enabled = false,						-- default state (not researched)
		ingredients =
		{
			{"iron-plate", 20},			
			{"steel-plate", 20},
			{"iron-gear-wheel", 10},				
			{"engine-unit", 15},
			{"copper-plate", 10}			-- Unique
		},
		result = "minigun-platform-mk1"
	}
})



--------------
--- ENTITY ---
--------------
-- Deep copy base data and create new one with custom parametres
local l_minigun_platform_mk1 = util.table.deepcopy(data.raw["cargo-wagon"]["cargo-wagon"])
l_minigun_platform_mk1.name = "minigun-platform-mk1"
l_minigun_platform_mk1.icon = "__Armored-train__/assets/minigun-platform-mk1/minigun-platform-mk1-icon.png"
l_minigun_platform_mk1.icon_size = 64
l_minigun_platform_mk1.icon_mipmaps = 4
-- Inventory
l_minigun_platform_mk1.inventory_size = 10
l_minigun_platform_mk1.minable = 
{
	mining_time = 0.5, 
	result = "minigun-platform-mk1"		
}
l_minigun_platform_mk1.max_health = 1000					-- Same as turret hp (Critical component or will be destroyed using old script)	
l_minigun_platform_mk1.weight = 2000
l_minigun_platform_mk1.energy_per_hit_point = 5
-- Animation
l_minigun_platform_mk1.pictures =
{
	layers =
	{
		{
			priority = "very-low",
			width = 256,
			height = 256,
			back_equals_front = false,    			--means to not rotate mirroring
			direction_count = 128, 					--means to add aditional frames
			allow_low_quality_rotation = true,
			line_length = 32,
			lines_per_file = 1,
			shift = {0.4, -1.25},
			filenames =
			{
				"__Armored-train__/assets/minigun-platform-mk1/sprites/platform-mk1-01.png",
				"__Armored-train__/assets/minigun-platform-mk1/sprites/platform-mk1-02.png",
				"__Armored-train__/assets/minigun-platform-mk1/sprites/platform-mk1-03.png",
				"__Armored-train__/assets/minigun-platform-mk1/sprites/platform-mk1-04.png"
			}
		}
	}
}
l_minigun_platform_mk1.horizontal_doors = nil
l_minigun_platform_mk1.vertical_doors = nil
-- Minimap representation
l_minigun_platform_mk1.minimap_representation =
{
	filename = "__Armored-train__/assets/minigun-platform-mk1/minigun-platform-mk1-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}
l_minigun_platform_mk1.selected_minimap_representation =
{
	filename = "__Armored-train__/assets/minigun-platform-mk1/minigun-platform-mk1-selected-minimap-representation.png",
	flags = {"icon"},
	size = {20, 40},
	scale = 0.5
}

-- Write result
data:extend
({
	l_minigun_platform_mk1
})

assets/rocket-turret-mk1/rocket-turret-mk1.lua
--------------
--- ENTITY ---
--------------
-- Deep copy base data and create new one with custom parametres
local l_rocket_turret_mk1 = util.table.deepcopy(data.raw["ammo-turret"]["gun-turret"])
l_rocket_turret_mk1.name = "rocket-turret-mk1"
-- Flags
l_rocket_turret_mk1.flags = 
{
	"player-creation",				-- Can draw enemies
	"placeable-off-grid", 
	"not-on-map",					-- Do not show on minimap

}
l_rocket_turret_mk1.minable = nil
-- HP 
l_rocket_turret_mk1.max_health = 1000			-- Same as platform hp (Critical component or will be destroyed using old script)
-- Resistance copy
local l_base_wagon_copy_param = util.table.deepcopy(data.raw["cargo-wagon"]["cargo-wagon"]["resistances"])	
l_rocket_turret_mk1.resistances = l_base_wagon_copy_param			-- Prevent faster damage from turret
-- Collision
l_rocket_turret_mk1.collision_box = {{-1, -1 }, {1, 1}}
l_rocket_turret_mk1.collision_mask = { "object-layer" }				-- Fix player stuck inside wagon forever
l_rocket_turret_mk1.selection_box = {{-1.25, -1.25 }, {1.25, 1.25}}
-- Rotation speed
local l_base_tank_copy_param = util.table.deepcopy(data.raw["car"]["tank"]["turret_rotation_speed"])		
l_rocket_turret_mk1.rotation_speed = l_base_tank_copy_param
-- Inventory
l_rocket_turret_mk1.inventory_size = 3
-- Ammo count copy
local l_base_ammo_copy_param = util.table.deepcopy(data.raw["ammo"]["rocket"]["stack_size"])	
l_rocket_turret_mk1.automated_ammo_count = l_base_ammo_copy_param * 2
-- Attack speed
l_rocket_turret_mk1.attacking_speed = l_rocket_turret_mk1.attacking_speed / 2

-- Animation functions
l_rocket_turret_mk1.gun_animation_render_layer = "higher-object-under"
local function rocket_turret_extension(inputs)
return
{
	filename = "__Armored-train__/assets/rocket-turret-mk1/sprites/rocket-turret-raising.png",
	priority = "medium",
	width = 88,
	height = 80,
	direction_count = 4,
	frame_count = inputs.frame_count or 5,
	line_length = inputs.line_length or 0,
	run_mode = inputs.run_mode or "forward",
	shift = util.by_pixel(0, -50),						-- Sprite offset in world
	scale = 1,											-- Sprite scale
	axially_symmetrical = false
	-- hr version missing
}
end
function rocket_turret_attack(inputs)
return
{
	layers =
	{
		{
			width = 88,
			height = 80,
			frame_count = inputs.frame_count or 2,
			axially_symmetrical = false,
			direction_count = 64,
			shift = util.by_pixel(0, -50),
			scale = 1,
			stripes =
			{
				{
					filename = "__Armored-train__/assets/rocket-turret-mk1/sprites/rocket-turret-shooting-1.png",
					width_in_frames = inputs.frame_count or 2,
					height_in_frames = 16
				},
				{
					filename = "__Armored-train__/assets/rocket-turret-mk1/sprites/rocket-turret-shooting-2.png",
					width_in_frames = inputs.frame_count or 2,
					height_in_frames = 16
				},
				{
					filename = "__Armored-train__/assets/rocket-turret-mk1/sprites/rocket-turret-shooting-3.png",
					width_in_frames = inputs.frame_count or 2,
					height_in_frames = 16
				},
				{
					filename = "__Armored-train__/assets/rocket-turret-mk1/sprites/rocket-turret-shooting-4.png",
					width_in_frames = inputs.frame_count or 2,
					height_in_frames = 16
				}
			}
			-- hr version missing and mask and shadow
		}
	}
}
end
-- Animation calls
l_rocket_turret_mk1.folded_animation =
{
	layers =
	{
		rocket_turret_extension{frame_count=1, line_length = 1},
		--gun_turret_extension_mask{frame_count=1, line_length = 1},
		--gun_turret_extension_shadow{frame_count=1, line_length = 1}
	}
}
l_rocket_turret_mk1.preparing_animation =
{
	layers =
	{
		rocket_turret_extension{},
		--gun_turret_extension_mask{},
		--gun_turret_extension_shadow{}
	}
}
-- Collision
l_rocket_turret_mk1.collision_box = {{-0.3, -0.3 }, {0.3, 0.3}} -- made turret smaller
l_rocket_turret_mk1.selection_box = {{-0.4, -0.4 }, {0.4, 0.4}} -- kept selection box slightly larger
l_rocket_turret_mk1.prepared_animation = rocket_turret_attack{frame_count=1}
l_rocket_turret_mk1.attacking_animation = rocket_turret_attack{}
l_rocket_turret_mk1.folding_animation =
{
	layers =
	{
		rocket_turret_extension{run_mode = "backward"},
		--gun_turret_extension_mask{run_mode = "backward"},
		--gun_turret_extension_shadow{run_mode = "backward"}
	}
}
-- Base picture
local blank_layers = 
{
	layers = 
	{
		{
			filename = 	"__Armored-train__/assets/minigun-turret-mk1/fakeTransparent.png",
			direction_count = 1,
			height = 16,
			width = 16
		}
	}
}
l_rocket_turret_mk1.base_picture = blank_layers

-- Attack parameters
local l_base_gun_copy = util.table.deepcopy(data.raw["gun"]["rocket-launcher"])
l_rocket_turret_mk1.attack_parameters = l_base_gun_copy.attack_parameters
l_rocket_turret_mk1.attack_parameters.projectile_center = {-0.17, -0.75}					-- Overwrite parameter
-- Range
l_rocket_turret_mk1.attack_parameters.min_range = 3.5										-- Overwrite parameter

-- Write result
data:extend
({
	l_rocket_turret_mk1
})